File: seach2dMatrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if matrix is None or len(matrix[0]) == 0: return False

        rows = len(matrix) - 1
        column = len(matrix[0]) - 1
        cindex, rindex = column, 0
        while cindex >= 0 and rindex < rows:
            if matrix[rindex][cindex] == target:
                return True
            elif matrix[rindex][cindex] > target:
                cindex -= 1
            else:
                rindex += 1

        return False


#time:-O(n)
# space:-O(1)
















if __name__ == '__main__':
    print(Solution().searchMatrix(nums=[1,1,1,2,2,3],k=2))

File: Seach2DMatrix.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        // check for null or empty array
        if(matrix == null || matrix.length == 0) return false;
        // intialize co-ordinates to the top right corner element of matrix
        int i = 0; int j = matrix[0].length - 1;
        // iterate through the matrix
        while(i>= 0 && j>=0 && i<matrix.length && j < matrix[0].length){
            // if target element is found then return true
            if(target == matrix[i][j]) return true;
            // eliminate right most column if the tagret is lesser than the element at the co-ordinates
            else if(target < matrix[i][j]) j--;
            // eliminare top most tow if the target is greater than the element at the co-ordinates
            else i++;
        }
        // return false if target element is not found in the matrix
        return false;
    }
}